### ISSUE TITLE
Ambiguous match found with ClassData and Decimal type

### ISSUE THREAD

**Issue by ardalis** [xunit/xunit#2298]
Probably related to #1771, if not duplicate.

This was working with version 2.3.1 but dependabot bumped it to 2.4.1 and the errors/failure started.

The code is open source here: https://github.com/ardalis/GuardClauses

The specific tests failing are here:
https://github.com/ardalis/GuardClauses/blob/master/test/GuardClauses.UnitTests/GuardAgainstOutOfRangeForInvalidInput.cs

This fails:

```
        public class CorrectClassData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { 20.0m, (Func<decimal, bool>)((x) => x > 10.0m) };
                yield return new object[] { 20.0, (Func<double, bool>)((x) => x > 10.0) };
            }
            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }
```

This succeeds:

```
        public class CorrectClassData : IEnumerable<object[]>
        {
            public IEnumerator<object[]> GetEnumerator()
            {
                //yield return new object[] { 20.0m, (Func<decimal, bool>)((x) => x > 10.0m) };
                yield return new object[] { 20.0, (Func<double, bool>)((x) => x > 10.0) };
            }
            IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
        }
```

**Comment by bradwilson** [comment]
I'll leave this open and verify once 1771 is fixed. Thanks!

**Comment by joaonunomota** [comment]
This issue was fixed in version 2.4.2 along with #1771. Tests from this scenario are included in https://github.com/xunit/xunit/issues/1771#issuecomment-1572977218 and a PR with working tests also exists in https://github.com/ardalis/GuardClauses/pull/276 (awaiting workflow approval at time of writing).

**Comment by bradwilson** [comment]
Thanks @joaonunomota!
